### PR TITLE
Feature/40 show legal links from registry

### LINKED ISSUE #40: Provide legal links in login/logout page from CES registry
There should be no longer just two legal URLs in the bottom in of the login or logout page but three, namely in this sequence:

1. Terms of Service
2. Imprint
3. Privacy Policy

These links must be extracted from the registry and may point anywhere.

If a registry value is missing or empty no respective link must be shown and with no default value.

All three links still must be translated in DE and EN

### PR DESCRIPTION
**Pull request by ppxl**
Fetch legal URLs from registry to show in bottom page fragment

The JSP fragment `bottom.jsp` is included in almost all pages that CAS
delivers. Previously, this bottom fragment contained two legal links to
imprint and the privacy policy.

This commit changes this behavior so 3 links from the registry can be
shown. Additional to the existing two links the new link sequence is now:

1. Terms of Service
2. Imprint
3. Privacy Policy

If one or more entries from the registry is missing then no links
will be shown. Entries not to be found are considered empty. Empty
entries are not rendered. Likewise, potentially superfluous link
delimiters may be subject of being not rendered.

Since registry access from the CAS is a rather expensive operation and
while the legal URLs are not supposed to change frequently these 3 links
are cached within the LegalLinkProducer bean.

This PR should resolve #40

### CODE REVIEW

**Comment by cesmarvin** [comment]
SonarQube analysis reported 1 issue
* ![CRITICAL][CRITICAL] 1 critical

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'

File: app/src/main/java/de/triology/cas/services/LegalLinkProducer.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * This class provides links from the etcd registry for the bottom fragment of every view that uses it. If one link is
+ * not set or available no respective link should be seen.
```
Comment (robertauer) [inline app/src/main/java/de/triology/cas/services/LegalLinkProducer.java]: "not set or not available..." or "not set or unavailable..." I would suggest

File: app/src/main/java/de/triology/cas/services/LegalLinkProducer.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    private static final Logger LOG = LoggerFactory.getLogger(LegalLinkProducer.class);
+    private static final String REGISTRY_KEY_TERMS_OF_SERVICE = "/config/cas/legal_urls/terms_of_service";
+    private static final String REGISTRY_KEY_IMPRINT = "/config/cas/legal_urls/imprint";
+    private static final String REGISTRY_KEY_PRIVACY_POLICY = "/config/cas/legal_urls/privacy_policy";
+    private static final String LINK_DELIMITER = "|";
+    private final RegistryEtcd registry;
+    /**
+     * This field provides a simple cache in order to reduce the strong etcd interaction latency.
+     * <h1>Note on cache invalidation</h1>
+     * Neither this class provide any means of cache invalidation mechanism because these data are
```
Comment (robertauer) [inline app/src/main/java/de/triology/cas/services/LegalLinkProducer.java]: I would suggest "This class does not provide any means...."

File: dogu.json [hunk 3]
```diff
@@ -48,4 +60,4 @@
     "Type": "tcp",
     "Port": 8080
   }
-}
+}TODO
```
Comment (robertauer) [inline dogu.json]: This needs to go I think ;)

File: dogu.json [hunk 4]
```diff
@@ -32,6 +32,18 @@
     "Name": "forgot_password_text",
     "Description": "Text which shall be shown if a user clicks 'Forgot Password'",
     "Optional": true
+  },{
+    "Name": "legal_urls/terms_of_service",
+    "Description": "This URL links to the Terms of Service and will be be the first link in bottom of login or logout pages. If empty the whole link will be omitted. If changed the CAS dogu must be restarted to take effect.",
```
Comment (robertauer) [inline dogu.json]: Maybe "... first link at the bottom of login..."

File: dogu.json [hunk 5]
```diff
@@ -32,6 +32,18 @@
...
+    "Optional": true
+  },{
+    "Name": "legal_urls/imprint",
+    "Description": "This URL links the Imprint and will be be the second link in bottom of login or logout pages. If empty the whole link will be omitted. If changed the CAS dogu must be restarted to take effect.",
```
Comment (robertauer) [inline dogu.json]: Maybe "This URL links to the Imprint..." and "... second link at the bottom of login..."

File: dogu.json [hunk 6]
```diff
@@ -32,6 +32,18 @@
...
+    "Optional": true
+  },{
+    "Name": "legal_urls/privacy_policy",
+    "Description": "This URL links the Privacy Policy and will be be the last link in bottom of login or logout pages. If empty the whole link will be omitted. If changed the CAS dogu must be restarted to take effect.",
```
Comment (robertauer) [inline dogu.json]: Maybe "This URL links to the Privacy..." and "... last link at the bottom of login..."

File: dogu.json (hunk 4 above)
Comment (ppxl) [reply dogu.json]: but you didnt catch the "be be" :rofl:

**Pull request merged**